pkg/deploymanager/subscription.go

```diff
@@ -0,0 +1,269 @@
+package deploymanager
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "github.com/operator-framework/api/pkg/operators/v1"
+	v1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+)
+
+type clusterObjects struct {
+	namespaces     []k8sv1.Namespace
+	operatorGroups []v1.OperatorGroup
+	catalogSources []v1alpha1.CatalogSource
+	subscriptions  []v1alpha1.Subscription
+}
+
+const marketplaceNamespace = "openshift-marketplace"
+
+// Generating the cluster objects.
+func (t *DeployManager) generateClusterObjects(lvmCatalogImage string, subscriptionChannel string) *clusterObjects {
+
+	co := &clusterObjects{}
+	label := make(map[string]string)
+	// Label required for monitoring this namespace
+	label["openshift.io/cluster-monitoring"] = "true"
+
+	// Namespaces
+	co.namespaces = append(co.namespaces, k8sv1.Namespace{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Namespace",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   InstallNamespace,
+			Labels: label,
+		},
+	})
+
+	// Operator Groups
+	lvmOG := v1.OperatorGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "openshift-storage-operatorgroup",
+			Namespace: InstallNamespace,
+		},
+		Spec: v1.OperatorGroupSpec{
+			TargetNamespaces: []string{InstallNamespace},
+		},
+	}
+	lvmOG.SetGroupVersionKind(schema.GroupVersionKind{Group: v1.SchemeGroupVersion.Group, Kind: "OperatorGroup", Version: v1.SchemeGroupVersion.Version})
+
+	co.operatorGroups = append(co.operatorGroups, lvmOG)
+
+	// Catalog Source
+	lvmCatalog := v1alpha1.CatalogSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "lvm-catalogsource",
+			Namespace: marketplaceNamespace,
+		},
+		Spec: v1alpha1.CatalogSourceSpec{
+			SourceType:  v1alpha1.SourceTypeGrpc,
+			Image:       "catalog-img",
+			DisplayName: "OpenShift Data Foundation",
+			Publisher:   "Red Hat",
+		},
+	}
+	lvmCatalog.SetGroupVersionKind(schema.GroupVersionKind{Group: v1alpha1.SchemeGroupVersion.Group, Kind: "CatalogSource", Version: v1alpha1.SchemeGroupVersion.Version})
+
+	co.catalogSources = append(co.catalogSources, lvmCatalog)
+
+	// Subscriptions
+	lvmSubscription := v1alpha1.Subscription{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "lvm-operator",
+			Namespace: InstallNamespace,
+		},
+		Spec: &v1alpha1.SubscriptionSpec{
+			Channel:                "alpha",
+			InstallPlanApproval:    "Automatic",
+			CatalogSource:          "lvm-catalogsource",
+			CatalogSourceNamespace: "openshift-marketplace",
+			StartingCSV:            "lvm-operator.v0.0.1",
```

Comment:
This field can be removed - it will work without it.

Reply:
Done